File: backend/src/haan/tools/memory_tools.py
"""
Persistent memory tools for Haan.ai agents.

Provides 3 tools mirroring the TypeScript memory-tools.ts:
- MemorySaveTool:   Save key-value information to long-term memory
- MemoryLoadTool:   Load a specific memory entry by key
- MemorySearchTool: Search memories by keyword across keys, content, and tags

Memory is stored as JSON files in the .haan/memory/ directory under the project root.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from crewai.tools import BaseTool as CrewAIBaseTool

from haan.utils.logger import logger
from haan.utils.path_utils import haan_sub_dir


def _get_memory_dir() -> Path:
    """Return the path to the memory storage directory."""
    return Path(haan_sub_dir("memory"))


def _sanitize_key(key: str) -> str:
    """Sanitize a memory key to a safe filename."""
    import re
    return re.sub(r"[^a-zA-Z0-9_-]", "_", key)


def _get_memory_path(key: str) -> Path:
    """Return the file path for a memory entry."""
    return _get_memory_dir() / f"{_sanitize_key(key)}.json"


def _load_entry(path: Path) -> dict[str, Any] | None:
    """Load a memory entry from a JSON file. Returns None on failure."""
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (FileNotFoundError, json.JSONDecodeError):
        return None


def _load_all_entries() -> list[dict[str, Any]]:
    """Load all memory entries from the memory directory."""
    memory_dir = _get_memory_dir()
    entries: list[dict[str, Any]] = []

    if not memory_dir.exists():
        return entries

    for file in sorted(memory_dir.glob("*.json")):
        entry = _load_entry(file)
        if entry:
            entries.append(entry)

    return entries


class MemorySaveTool(CrewAIBaseTool):
    """Save information to long-term memory."""

    name: str = "memory-save"
    description: str = (
        "Save information to long-term memory. Use this to remember important facts, "
        "decisions, context, or patterns. Specify a unique key and optional tags "
        "for later retrieval."
    )

    def _run(
        self,
        key: str,
        content: str,
        tags: list[str] | None = None,
    ) -> str:
        """
        Save a memory entry.

        Args:
            key: Unique key to identify this memory.
            content: Content to remember.
            tags: Optional tags for categorization and search.

        Returns:
            Success message indicating whether the entry was created or updated.
        """
        try:
            file_path = _get_memory_path(key)
            now = datetime.now(timezone.utc).isoformat()
            tag_list = tags or []

            # Check if entry already exists (for createdAt preservation)
            existing = _load_entry(file_path)

            entry = {
                "key": key,
                "content": content,
                "tags": tag_list,
                "createdAt": existing["createdAt"] if existing else now,
                "updatedAt": now,
            }

            file_path.write_text(json.dumps(entry, indent=2), encoding="utf-8")
            logger.info("memory", f"Saved memory: {key}")

            action = "Updated" if existing else "Saved"
            return f'{action} memory: "{key}" ({len(tag_list)} tags)'
        except Exception as e:
            return f"Error: Failed to save memory: {e}"


class MemoryLoadTool(CrewAIBaseTool):
    """Load a specific memory entry by key."""

    name: str = "memory-load"
    description: str = "Load a specific memory entry by key. Returns the content and metadata."

    def _run(self, key: str) -> str:
        """
        Load a memory entry.

        Args:
            key: Memory key to load.

        Returns:
            Memory content with metadata, or error if not found.
        """
        try:
            file_path = _get_memory_path(key)
            entry = _load_entry(file_path)

            if not entry:
                return f'Error: Memory not found: "{key}"'

            output = [
                f"Key: {entry['key']}",
                f"Tags: {', '.join(entry.get('tags', [])) or 'none'}",
                f"Created: {entry.get('createdAt', 'unknown')}",
                f"Updated: {entry.get('updatedAt', 'unknown')}",
                "",
                entry["content"],
            ]
            return "\n".join(output)
        except Exception as e:
            return f"Error: Failed to load memory: {e}"


class MemorySearchTool(CrewAIBaseTool):
    """Search memories by keyword."""

    name: str = "memory-search"
    description: str = (
        "Search memories by keyword. Searches across keys, content, and tags. "
        "Returns matching entries with content previews."
    )

    def _run(self, query: str) -> str:
        """
        Search memory entries.

        Args:
            query: Search query string.

        Returns:
            Matching memory entries with previews.
        """
        try:
            query_lower = query.lower()
            entries = _load_all_entries()

            matches = [
                e
                for e in entries
                if query_lower
                in " ".join([e["key"], e["content"]] + e.get("tags", [])).lower()
            ]

            if not matches:
                return f'No memories found matching: "{query}"'

            output_parts: list[str] = []
            for entry in matches:
                content = entry["content"]
                preview = content[:100] + "..." if len(content) > 100 else content
                tags = ", ".join(entry.get("tags", [])) or "none"
                output_parts.append(f"[{entry['key']}] (tags: {tags})\n  {preview}")

            return "\n\n".join(output_parts) + f"\n\n({len(matches)} matches)"
        except Exception as e:
            return f"Error: Memory search failed: {e}"

File: src/tools/memory-tools.ts
import fs from 'fs';
import path from 'path';
import { BaseTool, ToolResult } from './types.js';
import { haanSubDir } from '../utils/path-utils.js';
import { logger } from '../utils/logger.js';

interface MemoryEntry {
  key: string;
  content: string;
  tags: string[];
  createdAt: string;
  updatedAt: string;
}

function getMemoryDir(): string {
  return haanSubDir('memory');
}

function getMemoryFilePath(key: string): string {
  // Sanitize key to safe filename
  const safe = key.replace(/[^a-zA-Z0-9_-]/g, '_');
  return path.join(getMemoryDir(), `${safe}.json`);
}

function loadEntry(filePath: string): MemoryEntry | null {
  try {
    const raw = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(raw) as MemoryEntry;
  } catch {
    return null;
  }
}

function loadAllEntries(): MemoryEntry[] {
  const dir = getMemoryDir();
  let files: string[];
  try {
    files = fs.readdirSync(dir).filter((f) => f.endsWith('.json'));
  } catch {
    return [];
  }

  const entries: MemoryEntry[] = [];
  for (const file of files) {
    const entry = loadEntry(path.join(dir, file));
    if (entry) entries.push(entry);
  }
  return entries;
}

export class MemorySaveTool extends BaseTool {
  name = 'memory-save';
  description = 'Save information to long-term memory. Use this to remember important facts, decisions, context, or patterns.';
  parameters = {
    type: 'object' as const,
    properties: {
      key: { type: 'string', description: 'Unique key to identify this memory' },
      content: { type: 'string', description: 'Content to remember' },
      tags: {
        type: 'array',
        items: { type: 'string' },
        description: 'Tags for categorization and search',
      },
    },
    required: ['key', 'content'],
  };

  async execute(params: Record<string, unknown>): Promise<ToolResult> {
    try {
      const key = params.key as string;
      const content = params.content as string;
      const tags = (params.tags as string[]) || [];

      const filePath = getMemoryFilePath(key);
      const now = new Date().toISOString();

      const existing = loadEntry(filePath);
      const entry: MemoryEntry = {
        key,
        content,
        tags,
        createdAt: existing?.createdAt || now,
        updatedAt: now,
      };

      fs.writeFileSync(filePath, JSON.stringify(entry, null, 2), 'utf-8');
      logger.info('memory-save', `Saved memory: ${key}`);

      const action = existing ? 'Updated' : 'Saved';
      return this.success(`${action} memory: "${key}" (${tags.length} tags)`, {
        key,
        tags,
        action: action.toLowerCase(),
      });
    } catch (err) {
      return this.error(`Failed to save memory: ${(err as Error).message}`);
    }
  }
}

export class MemoryLoadTool extends BaseTool {
  name = 'memory-load';
  description = 'Load a specific memory by key.';
  parameters = {
    type: 'object' as const,
    properties: {
      key: { type: 'string', description: 'Memory key to load' },
    },
    required: ['key'],
  };

  async execute(params: Record<string, unknown>): Promise<ToolResult> {
    try {
      const key = params.key as string;
      const filePath = getMemoryFilePath(key);

      const entry = loadEntry(filePath);
      if (!entry) {
        return this.error(`Memory not found: "${key}"`);
      }

      const output = [
        `Key: ${entry.key}`,
        `Tags: ${entry.tags.join(', ') || 'none'}`,
        `Created: ${entry.createdAt}`,
        `Updated: ${entry.updatedAt}`,
        '',
        entry.content,
      ].join('\n');

      return this.success(output, { key: entry.key, tags: entry.tags });
    } catch (err) {
      return this.error(`Failed to load memory: ${(err as Error).message}`);
    }
  }
}

export class MemorySearchTool extends BaseTool {
  name = 'memory-search';
  description = 'Search memories by keyword. Searches keys, content, and tags.';
  parameters = {
    type: 'object' as const,
    properties: {
      query: { type: 'string', description: 'Search query' },
    },
    required: ['query'],
  };

  async execute(params: Record<string, unknown>): Promise<ToolResult> {
    try {
      const query = (params.query as string).toLowerCase();
      const entries = loadAllEntries();

      const matches = entries.filter((e) => {
        const searchable = [e.key, e.content, ...e.tags].join(' ').toLowerCase();
        return searchable.includes(query);
      });

      if (matches.length === 0) {
        return this.success(`No memories found matching: "${params.query}"`);
      }

      const output = matches
        .map((e) => {
          const preview = e.content.length > 100 ? e.content.slice(0, 100) + '...' : e.content;
          return `[${e.key}] (tags: ${e.tags.join(', ') || 'none'})\n  ${preview}`;
        })
        .join('\n\n');

      return this.success(output, { matchCount: matches.length });
    } catch (err) {
      return this.error(`Memory search failed: ${(err as Error).message}`);
    }
  }
}
